internal/storage/bucket_handle.go

```diff
@@ -481,3 +485,59 @@ func (b *bucketHandle) getStorageLayout() (*controlpb.StorageLayout, error) {
 func isStorageConditionsNotEmpty(conditions storage.Conditions) bool {
 	return conditions != (storage.Conditions{})
 }
+
+func (b *bucketHandle) ParallelDownloadToFile(ctx context.Context, req *gcs.ParallelDownloadToFileRequest) error {
+
+	if b.downloader == nil {
+		return fmt.Errorf("parallel download not suppported")
+	}
+
+	if req.PartSize == 0 {
+		return fmt.Errorf("invalid part-size")
+	}
+
+	size := req.Range.Limit - req.Range.Start
+	numParts := (size-1)/req.PartSize + 1
+
+	wg := sync.WaitGroup{}
+
+	callback := func(out *transfermanager.DownloadOutput) {
```

Comment:
@tritone What should be the right way to return the error on the first encounter and cancel the other worker?

Reply:
I think it would make sense to launch a sentinel goroutine that can monitor for errors via the callback, then cancel the context for the downloads if something errors. I will add this to my paste example.

Reply:
I updated the paste with this, PTAL